pipewire: update to 0.3.31.

paper42 opened the pull request:
#### General
- [ ] This is a new package and it conforms to the [quality requirements](https://github.com/void-linux/void-packages/blob/master/Manual.md#quality-requirements)

#### Have the results of the proposed changes been tested?
- [x] I use the packages affected by the proposed changes on a regular basis and confirm this PR works for me
- [ ] I generally don't use the affected packages but briefly tested this PR

0.3.29 introduces `module-rt` which is supposed to replace rtkit on systems without it. I was able to make this work on my glibc system (replacing module-rtkit with module-rt in the configs and setting CAP_SYS_NICE on the binaries), but not on musl.

ping @ericonr, @st3r4g

ericonr commented:
Testing locally now.

st3r4g commented:
> but not on musl.

It doesn't work because the new module still uses `sched_setscheduler`, which is not implemented in musl because the Linux API is per thread, not per process as required by the POSIX API.
I want to try if switching to `pthread_setschedparam(pthread_self(), ...)` makes it work on musl.

ericonr commented:
Btw, should we introduce a build option or document this in some way? I assume the `CAP_SYS_NICE` requirement is for the `pipewire` binary?

st3r4g commented:
Some tentative runit scripts for system-wide pipewire & pipewire-pulse.
`/etc/sv/pipewire`:
```
#!/bin/sh
mkdir -p /run/pipewire
umask 000
PIPEWIRE_RUNTIME_DIR=/run/pipewire pipewire
```
`/etc/sv/pipewire-pulse`:
```
#!/bin/sh
mkdir -p /run/pulse
umask 000
PULSE_RUNTIME_PATH=/run pipewire-pulse
```
Audio works for all users with these for me, but for some reason I get a connection failure to system dbus... screensharing works too.

st3r4g commented:
> Btw, should we introduce a build option or document this in some way? I assume the `CAP_SYS_NICE` requirement is for the `pipewire` binary?

If I understand correctly, `module-rt` cannot be an exact replacement to `module-rtkit`, because all the clients would need`CAP_SYS_NICE` too to boost their own processing thread. I wonder how jack handles this...

paper42 commented:
@st3r4g I added your suggestions to the package, could you test it? I am sorry for taking credit for this PR even though you did most of the work.

st3r4g commented on srcpkgs/pipewire/template:
```diff
@@ -18,7 +18,9 @@ license="MIT"
 distfiles="https://gitlab.freedesktop.org/pipewire/pipewire/-/archive/${version}/pipewire-${version}.tar.gz"
-checksum=1d9271e121a5049aef379e9bb7c50524faa6f971e668806637d7b9df1b7cab88
+checksum=09a002a0846e168e8dc61b3aa579d38885aef32ac971351d5d3cde20bcd1759b
+make_dirs="/var/lib/pipewire 0755 root root"
```
this needs to be owned by user pipewire

st3r4g commented on srcpkgs/pipewire/files/pipewire/run:
```diff
@@ -0,0 +1,5 @@
+mkdir -p /run/pipewire
+chown pipewire:pipewire /run/pipewire
+umask 002
+PIPEWIRE_RUNTIME_DIR=/run/pipewire XDG_CONFIG_HOME=/var/lib chpst -u pipewire:pipewire:audio:video pipewire
```
forgot exec here

st3r4g commented on srcpkgs/pipewire/files/pipewire/run:
```diff
@@ -0,0 +1,5 @@
+#!/bin/sh
+mkdir -p /run/pipewire
+chown pipewire:pipewire /run/pipewire
```
maybe we could do this in a single command via `install`, (see dbus service). But then we also need to assign a fixed id to user pipewire.

cinerea0 replied:
I'm pretty sure it should be `_pipewire`.

heliocat replied:
`install -o` can take both UIDs and names, ditto `-g`.

st3r4g replied:
that's true, my bad

paper42 replied:
> I'm pretty sure it should be `_pipewire`.

I am not sure, I think _pipewire would be used for a group which users do not have to interact with (users should never be a part of it). Names without the leading _ like wheel are often used for groups which users might want to be a part of. If we agree on this, then the pipewire group shouldn't have a leading _ in its name.

ericonr replied:
I believe the namespacing is more about avoiding conflicts than specifying what a user should or shouldn't touch. Unless it's a common name (like http or plugdev), creating a user name without a leading underscore should always be harmless.

st3r4g commented on srcpkgs/pipewire/files/pipewire-pulse/run:
```diff
@@ -0,0 +1,5 @@
+#!/bin/sh
+mkdir -p /run/pulse
+chown pipewire:pipewire /run/pulse
```
this could mess with system pulsaudio, in case someone has enabled both by mistake. It should probably be guarded

paper42 replied:
This will need adding a check script to pulseaudio, right?

st3r4g replied:
actually it might be fine now that there's no chown anymore. the pipewire service will just fail, which is ok.

st3r4g replied:
It fails with "Address already in use", it's ok as is

st3r4g reviewed (commented):
> @st3r4g I added your suggestions to the package, could you test it? I am sorry for taking credit for this PR even though you did most of the work.

No problem, I don't mind. These are collaborative efforts. I will test it on another pc with runit, for now I just translated my execline scripts. But it would be good if someone could check that bluetooth works, which I don't use. And I've not yet figured out the dbus error...

paper42 commented:
> > @st3r4g I added your suggestions to the package, could you test it? I am sorry for taking credit for this PR even though you did most of the work.
> 
> No problem, I don't mind. These are collaborative efforts. I will test it on another pc with runit, for now I just translated my execline scripts. But it would be good if someone could check that bluetooth works, which I don't use. And I've not yet figured out the dbus error...

I will try bluetooth tomorrow, I will have to reboot to a runit system. Do you have s6 services somewhere?

st3r4g commented:
> Do you have s6 services somewhere?

I'm using these s6-rc definitions
[pipewire-s6-rc.tar.gz](https://github.com/void-linux/void-packages/files/6683095/pipewire-s6-rc.tar.gz)

st3r4g commented:
About the "Failed to connect to system bus" error, it's a bug in pipewire logging: it's failing to connect to the *session* bus, which of course is not there.

st3r4g commented on srcpkgs/pipewire/files/pipewire/run:
```diff
@@ -0,0 +1,4 @@
+#!/bin/sh
+! [ -d /run/pipewire ] && install -m 755 -g pipewire -o pipewire -d /run/pipewire
+umask 002
+PIPEWIRE_RUNTIME_DIR=/run/pipewire XDG_CONFIG_HOME=/var/lib exec chpst -u pipewire:pipewire pipewire
```
needs pipewire:pipewire:**audio:video** too to work

st3r4g replied:
Could you also split this line into multiple lines with `export ...` to make it more readable? same in pipewire-pulse

paper42 replied:
> needs pipewire:pipewire:**audio:video** too to have the permissions to open hardware devices

done, this is not needed in the pipewire-pulse service right?

st3r4g replied:
> done, this is not needed in the pipewire-pulse service right?

right

st3r4g commented:
> I wonder how jack handles this...

I think I got it. `/etc/security/limits.d/jack.conf` simply allows unprivileged process to set any RT priority.

paper42 commented:
sorry for the delay, in the meantime, a new upstream release got tagged. This release adds a new testing framework that makes some tests segfault.

* added libusb to makedepends, it should help with recognizing bluetooth device capabilities (0.3.31)
* removed the rttime patch as it is upstreamed now (0.3.31)
* changed the maintainer to @st3r4g

st3r4g commented:
```
Starting program: /builddir/pipewire-0.3.31/build/test/test-pwtest 
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/usr/lib64/libthread_db.so.1".
[Detaching after fork from child process 9045]

Program received signal SIGSEGV, Segmentation fault.
0x00007ffff7dae371 in __GI__IO_fwrite (buf=buf@entry=0x555555559053, size=size@entry=1, count=count@entry=7, fp=fp@entry=0x0) at iofwrite.c:37
37	iofwrite.c: No such file or directory.
(gdb) bt
#0  0x00007ffff7dae371 in __GI__IO_fwrite (buf=buf@entry=0x555555559053, size=size@entry=1, count=count@entry=7, fp=fp@entry=0x0) at iofwrite.c:37
#1  0x00005555555581f2 in fprintf (__fmt=0x555555559053 "pwtest\n", __stream=0x0) at /usr/include/bits/stdio2.h:103
#2  make_xdg_runtime_dir () at ../test/pwtest.c:1077
#3  0x00005555555566e2 in main (argc=<optimized out>, argv=<optimized out>) at ../test/pwtest.c:1271
(gdb) f 2
#2  make_xdg_runtime_dir () at ../test/pwtest.c:1077
1077		fprintf(fp, "pwtest\n");
(gdb) l
1072		/* Marker file to avoid removing a random directory during cleanup */
1073		r = spa_scnprintf(path, sizeof(path), "%s/pwtest.dir", dir);
1074		spa_assert((size_t)r == strlen(dir) + 11);
1075		fp = fopen(path, "w");
1076		spa_assert(fp);
1077		fprintf(fp, "pwtest\n");
1078		fclose(fp);
1079	
1080		return dir;
1081	}
(gdb) p fp
$1 = (FILE *) 0x0
(gdb) p path
$2 = "/tmp/pwtest-11:35-XXXXXX/pwtest.dir", '\000' <repeats 230 times>...
(gdb) p dir
$3 = 0x55555555da20 "/tmp/pwtest-11:35-XXXXXX"
```
looks like `spa_assert` is not working?

paper42 commented:
@st3r4g I enabled assertions only when xbps-src runs check, but this is not really ideal. If the tests depend on assertions only in the tests/ directory (?), upstream could probably force enable assertions there. Would you like to create an issue or a PR?

ericonr commented:
Given how meson works, it stands to reason that they expect assertions to be enabled for the final program as well. I would rather have them enabled at all times instead of building the package differently if tests are enabled or not.

st3r4g commented:
> Given how meson works, it stands to reason that they expect assertions to be enabled for the final program as well.

Meson doesn't disable assertions by default in release builds, right? Then yes, I agree.

ericonr commented on srcpkgs/pipewire/template:
```diff
@@ -1,24 +1,26 @@
-checksum=1d9271e121a5049aef379e9bb7c50524faa6f971e668806637d7b9df1b7cab88
+checksum=2fec0eb47dbfcad51fb8bb3d59c26fe57e09204d63c9d2776070dbdc08aaaaae
+make_dirs="/var/lib/pipewire 0755 pipewire pipewire"
+system_accounts="pipewire"
```
Why is this still going with `pipewire`? Packages should ideally stick to the `_*` namespace.

st3r4g replied:
upstream uses the `pipewire` user name, but I have no problem adding `_` to be consistent with void's policy

ericonr replied:
`_seatd` wasn't an upstream recommendation either... I think using `_pipewire` is more correct.

paper42 replied:
sorry, fixed

ericonr reviewed (commented):
For completeness, the commit message could mention that libusb was added to improve bluetooth device detection; and I think since the package already has a `README.voidlinux` file, we could document the service as "experimental", so making breaking changes in a new release is possible.

ericonr commented on srcpkgs/pipewire/files/pipewire/run:
```diff
@@ -0,0 +1,6 @@
+#!/bin/sh
+! [ -d /run/pipewire ] && install -m 755 -g pipewire -o pipewire -d /run/pipewire
+umask 002
+export PIPEWIRE_RUNTIME_DIR=/run/pipewire
+export XDG_CONFIG_HOME=/var/lib
```
How do you feel about making `PIPEWIRE_RUNTIME_DIR` configurable? Does that make sense?

st3r4g replied:
I think there's no reason to make it configurable. It's the same in [upstream's systemd unit](https://gitlab.freedesktop.org/pipewire/pipewire/-/blob/master/src/daemon/systemd/system/pipewire.service.in) (`%t` expands to /run I guess)

paper42 commented:
> For completeness, the commit message could mention that libusb was added to improve bluetooth device detection; and I think since the package already has a `README.voidlinux` file, we could document the service as "experimental", so making breaking changes in a new release is possible.

I also mentioned that most users want to start pipewire{,-pulse} as a user, because I think there will be a lot of people trying to use the service without reading the handbook. I also mentioned it inside the service scripts.

dkwo commented:
With the update from 28->31, pipewire does not show up in alsamixer anymore, it says `This sound device does not have any controls.`, while before it was possible to raise volume/mute/etc using amixer commands.
Is this intentional or a bug?

ericonr commented:
There were some ALSA changes, I could see a bug slipping in. The changelog doesn't make it seem intentional.

st3r4g commented:
@dkwo looks like https://gitlab.freedesktop.org/pipewire/pipewire/-/issues/1384
should be fixed in next release

dkwo commented:
Great, thanks for pointing out.
So I'll just wait :)
